### PR TITLE
FINERACT-1724: Ability to customize the EntityManagerFactory from custom modules

### PR DESCRIPTION
**Pull request by galovics**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java [hunk 1]
```diff
@@ -70,29 +77,48 @@ public JPAConfig(RoutingDataSource dataSource, JpaProperties properties, ObjectP
     @DependsOn("tenantDatabaseUpgradeService")
     public LocalContainerEntityManagerFactoryBean entityManagerFactory(EntityManagerFactoryBuilder factoryBuilder) {
         Map<String, Object> vendorProperties = getVendorProperties();
-        customizeVendorProperties(vendorProperties);
-        return factoryBuilder.dataSource(getDataSource()).properties(vendorProperties).persistenceUnit("jpa-pu")
-                .packages("org.apache.fineract").jta(false).build();
+        String[] packagesToScan = getPackagesToScan();
+        return factoryBuilder.dataSource(getDataSource()).properties(vendorProperties).persistenceUnit("jpa-pu").packages(packagesToScan)
```
Comment (vidakovic) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java:81]: Do we need the customizer? Wouldn't it be "easier" with a simple configuration property to pass the packages as an array of strings here?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java [hunk 2]
```diff
@@ -70,29 +77,48 @@ public JPAConfig(RoutingDataSource dataSource, JpaProperties properties, ObjectP
...
+                .jta(false).build();
+    }
+
+    @Override
+    protected Map<String, Object> getVendorProperties() {
+        Map<String, Object> vendorProperties = new HashMap<>();
+        vendorProperties.put(PersistenceUnitProperties.WEAVING, "static");
+        vendorProperties.put(PersistenceUnitProperties.PERSISTENCE_CONTEXT_CLOSE_ON_COMMIT, "true");
+        vendorProperties.put(PersistenceUnitProperties.CACHE_SHARED_DEFAULT, "false");
+        emFactoryCustomizers.forEach(c -> vendorProperties.putAll(c.additionalVendorProperties()));
```
Comment (vidakovic) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java:91]: Same here... adding this properties (with defaults) to application.properties would be easier?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java (hunk 1 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java:81]: Not necessarily. If you look at the JpaBaseConfiguration class - provided by Spring - they do programmatic resolution of necessary packages. I can imagine a similar situation where you have some logic to figure out which packages are important.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java (hunk 2 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JPAConfig.java:91]: Defaults, might be but I don't want to have a fragmented configuration codebase for JPA. either everything sits here or in app props.

Plus, these customizers will be used by custom modules/external modules where the application.properties resolution (in addition to Fineract's) is not solved. This is similar how Spring expands its configuration capabilities tbh.

**Review by vidakovic — APPROVED** [review]
LGTM

**Pull request merged**